### PR TITLE
[SPARK-13205] [SQL] SQL Generation Support for Self Join

### PR DESCRIPTION
**Pull request by gatorsmile**
This PR addresses two issues:
- Self join does not work in SQL Generation
- When creating new instances for `LogicalRelation`, `metastoreTableIdentifier` is lost.

@liancheng Could you please review the code changes? Thank you!

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 1]
```diff
@@ -142,7 +142,15 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
       Some(s"`$database`.`$table`")
 
     case Subquery(alias, child) =>
-      toSQL(child).map(childSQL => s"($childSQL) AS $alias")
+      toSQL(child).map( childSQL =>
+        child match {
+          // Parentheses is not used for persisted data source relations
+          // e.g., select x.c1 from (t1) as x inner join (t1) as y on x.c1 = y.c1
+          case Subquery(_, l: LogicalRelation) =>
+            s"$childSQL AS $alias"
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Should we also handle `MetastoreRelation` here?

**Comment by liancheng** [comment]
One minor issue, overall looks good. Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Yeah, let me do it now. Thanks!

**Comment by liancheng** [comment]
LGTM, merged to master. Thanks!

**Pull request closed without merging**